In [1]:
import pandas as pd

# Question 1: 

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 

- Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 
- What metric would you report for this dataset?
- What is its value?


# Exploration:

In [2]:
df = pd.read_csv('2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv')
df.head(3)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56

In [3]:
df['total_items'].value_counts()

2       1832
1       1830
3        941
4        293
5         77
2000      17
6          9
8          1
Name: total_items, dtype: int64

Looking at the total items in each order it would appear that there are a handful of bulk orders that are greatly skewing the mean.

In [4]:
df.loc[df['total_items'] == 2000]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  
15    2017-03-07 4:00:00  
60    2017-03-04 4:00:00  
520   2017-03-02 4:00:00  
1104  2017-03-24 4:00:00  
1362  2017-03-15 4:00:00  
1436  2017-03-11 4:00:00  
1562  2017-03-19 4:00:00  
1602  2017-03-17 4:00:00  
2153  2017-03-12 4:00:00  
2297  2017-03-07 4:00:00  
2835  2017-03-28 4:00:00  
2969  2017-03-28 4:00:00  
3332  2017-03-24 4:00:00  
4056  2017-03-28 4:00:00  
4646  2017-03-02 4:00:00  
4868  2017-03-22 4:00:00  
4882  2017-03-25 4:00:00

In [5]:
len(df.loc[(df['shop_id'] == 42) & (df['total_items']!= 2000)])

34

The bulk orders are all coming from the stame shop_id, however the shop does also make non bulk sales.

In [6]:
df['order_amount'].loc[df['total_items'] != 2000].mean()

754.0919125025085

Seperating the bulk order from the data, the AOV still seems unusually high

In [7]:
item_cost = (df['order_amount'] / df['total_items'])
item_cost.mean()

387.7428

Looking at the average item cost, it would appear to also be higher than expected

In [8]:
df.loc[item_cost > item_cost.mean()]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1056      1057       78      800         25725            1          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1259      1260       78      775         77175            3    credit_card   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1452      1453       78      812         25725            1    credit_card   
1529      1530       78      810         51450            2           cash   
2270      2271       78      855         25725            1    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2773      2774       78      890         25725            1           cash   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2906      2907       78      817         77175            3          debit   
2922      2923       78      740         25725            1          debit   
3085      3086       78      910         25725            1           cash   
3101      3102       78      855         51450            2    credit_card   
3151      3152       78      745         25725            1    credit_card   
3167      3168       78      927         51450            2           cash   
3403      3404       78      928         77175            3          debit   
3440      3441       78      982         25725            1          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
3780      3781       78      889         25725            1           cash   
4040      4041       78      852         25725            1           cash   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4505      4506       78      866         25725            1          debit   
4584      4585       78      997         25725            1           cash   
4715      4716       78      818         77175            3          debit   
4918      4919       78      823         25725            1           cash   

               created_at  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
511    2017-03-09 7:23:14  
617   2017-03-18 11:18:42  
691   2017-03-27 22:51:43  
1056  2017-03-15 10:16:45  
1193  2017-03-16 16:38:26  
1204  2017-03-17 22:32:21  
1259   2017-03-27 9:27:20  
1384  2017-03-17 16:38:06  
141

In [9]:
df.loc[(df['shop_id'] == 78) & (item_cost != 25725)]

Empty DataFrame
Columns: [order_id, shop_id, user_id, order_amount, total_items, payment_method, created_at]
Index: []

Seperating the records with a higher than average item cost gives a very clear explaination of why - the items being sold at shop #78 are being sold at $25,725 each which is a very clear outlier and would also be skewing the AOV to be larger than expected.

In [10]:
naive_average = df['order_amount'].mean()
average_minus_bulk = df['order_amount'].loc[df['total_items'] < 100].mean()
average_minus_shop78 = df['order_amount'].loc[df['shop_id'] != 78].mean()
average_minus_bulk_and_shop78 =  df['order_amount'].loc[(df['shop_id'] != 78) & (df['total_items'] < 100)].mean()
print(f"""
      Naive Average:                        {naive_average}
      Average without bulk purchases:       {average_minus_bulk}
      Average without shop #78:             {average_minus_shop78}
      Average without bulk and shop #78:    {average_minus_bulk_and_shop78}
      """)


      Naive Average:                        3145.128
      Average without bulk purchases:       754.0919125025085
      Average without shop #78:             2717.3677836092047
      Average without bulk and shop #78:    302.58051448247926
      


# Thoughts:

Exploring the data I discovered two clear outliers that were the result of the higher than expected AOV.

The first is bulk purchases.  I would suggest filtering these out from the results as the average order value for individual customer purchases should probably be considered seperately from those of bulk purchases. 

The second is an extremely expensive sneaker at shop 78. I would also suggest filtering these from the results as they would appear to be a luxury item in their own category.

Both of these outliers are each respectively unique to a single store, and it would also be worth while speaking with someone at these stores to ensure this data is accurate and not the result of an error.  I would also want to consult with someone in the buisness department before implementing any thresholds that determine whether a purchase is bulk or not, or if an item is of high enough value to be in it's own category.

# Final Answer:

I would report an AOV of $302.58 as it gives the best impression of orders being made at the typical sneaker store.  However I would also note the outliers that had been removed from that calculation and bring attention to the signifigance that they hold in their own right.



<br></br>


# Question 2: 
- *How many orders were shipped by Speedy Express in total?*
```
SELECT COUNT(*) FROM Orders
    JOIN Shippers ON Shippers.ShipperID = Orders.ShipperID
    WHERE Shippers.ShipperName is 'Speedy Express';
```

52 orders were shipped by speedy express

<br/><br/>

- *What is the last name of the employee with the most orders?*
```
SELECT LastName from Employees
    JOIN Orders ON Employees.EmployeeID = Orders.EmployeeID
    GROUP BY Employees.EmployeeID
    ORDER BY COUNT(Orders.OrderID) DESC
    LIMIT 1;
```
Peacock is the last name of the employee with the most sales

<br/><br/>

 - What product was ordered the most by customers in Germany?
 ```
SELECT ProductName FROM Products
	JOIN OrderDetails ON Products.ProductID = OrderDetails.ProductID
    JOIN Orders ON OrderDetails.OrderID = Orders.OrderID
    JOIN Customers ON Orders.CustomerID = Customers.CustomerID
    	WHERE Customers.Country IS 'Germany'
    GROUP BY ProductName
    ORDER BY COUNT(Orders.OrderID) DESC
    LIMIT 1;
```
Gorgonzola Telino is the product ordered most by customers in Germany
